src/plugins/lua/gpt.lua

```diff
@@ -0,0 +1,336 @@
+  # Timeout for requests
+  timeout = 10s;
+  # Prompt for the model (use default if not set)
+  prompt = "xxx";
+  # Custom condition (lua function)
+  condition = "xxx";
+  # Autolearn if gpt classified
+  autolearn = true;
+  # Reply conversion (lua code)
+  reply_conversion = "xxx";
+  # URL for the API
+  url = "https://api.openai.com/v1/chat/completions";
+}
+  ]])
+  return
+end
+
+local lua_util = require "lua_util"
+local rspamd_http = require "rspamd_http"
+local rspamd_logger = require "rspamd_logger"
+local ucl = require "ucl"
+
+-- Exclude checks if one of those is found
+local default_symbols_to_except = {
+  'BAYES_SPAM', -- We already know that it is a spam, so we can safely skip it, but no same logic for HAM!
+  'WHITELIST_SPF',
+  'WHITELIST_DKIM',
+  'WHITELIST_DMARC',
+  'FUZZY_DENIED',
+}
+
+local settings = {
+  type = 'openai',
+  api_key = nil,
+  model = 'gpt-3.5-turbo',
+  max_tokens = 1000,
+  temperature = 0.7,
+  top_p = 0.9,
+  timeout = 10,
+  prompt = nil,
+  condition = nil,
+  autolearn = false,
+  url = 'https://api.openai.com/v1/chat/completions',
+  symbols_to_except = default_symbols_to_except,
+}
+
+local function default_condition(task)
+  -- Check result
+  -- 1) Skip passthrough
+  -- 2) Skip already decided as spam
+  -- 3) Skip already decided as ham
+  local result = task:get_metric_result()
+  if result then
+    if result.passthrough then
+      return false, 'passthrough'
+    end
+    local score = result.score
+    local action = result.action
+
+    if action == 'reject' and result.npositive > 1 then
+      return true, 'already decided as spam'
+    end
+
+    if action == 'no action' and score < 0 then
+      return true, 'negative score, already decided as ham'
+    end
+  end
+  -- We also exclude some symbols
+  for _, s in ipairs(settings.symbols_to_except) do
+    if task:has_symbol(s) then
+      return false, 'skip as "' .. s .. '" is found'
+    end
+  end
+
+  -- Check if we have text at all
+  local mp = task:get_parts() or {}
+  local sel_part
+  for _, mime_part in ipairs(mp) do
+    if mime_part:is_text() then
+      local part = mime_part:get_text()
+      if part:is_html() then
+        -- We prefer html content
+        sel_part = part
+      elseif not sel_part then
+        sel_part = part
+      end
+    end
+  end
+
+  if not sel_part then
+    return false, 'no text part found'
+  end
+
+  -- Check limits and size sanity
+  local nwords = sel_part:get_words_count()
+
+  if nwords < 5 then
+    return false, 'less than 5 words'
+  end
+
+  if nwords > settings.max_tokens then
+    -- We need to truncate words (sometimes get_words_count returns a different number comparing to `get_words`)
+    local words = sel_part:get_words('norm')
+    nwords = #words
+    if nwords > settings.max_tokens then
+      -- Trim something that does not fit
+      for i = nwords, settings.max_tokens, -1 do
+        rawset(words, i, nil)
+      end
+      return true, table.concat(words, ' ')
```

Comment:
How about `table.concat(words, ' ', 1, math.min(settings.max_tokens, nwords))`

Reply:
That looks better, thanks.